File: argon-plugin/src/main/scala/dev/argon/plugin/tube/SerializedTube.scala
package dev.argon.plugin.tube

import dev.argon.tube.*
import dev.argon.io.FileSystemResource
import zio.*

trait SerializedTube[R, E] {
  def version: ZIO[R, E, TubeFormatVersion]

  def metadata: ZIO[R, E, Metadata]

  def getResource(id: String): ZIO[R, E, FileSystemResource[R, E]]

  def getModule(modulePath: ModulePath): ZIO[R, E, ModuleDefinition]
  def getClass(id: BigInt): ZIO[R, E, ClassDefinition]
  def getTrait(id: BigInt): ZIO[R, E, TraitDefinition]
  def getFunction(id: BigInt): ZIO[R, E, FunctionDefinition]
  def getMethod(id: BigInt): ZIO[R, E, MethodDefinition]
  def getClassConstructor(id: BigInt): ZIO[R, E, ClassConstructorDefinition]
}

File: plugin/api/java/src/main/java/dev/argon/plugin/api/tube/SerializedTube.java
package dev.argon.plugin.api.tube;

import dev.argon.plugin.api.FileSystemResource;
import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.math.BigInteger;

public interface SerializedTube<E extends Throwable> extends AutoCloseable {
    @NotNull TubeFormatVersion version() throws E, IOException, InterruptedException;
    
    @NotNull Metadata metadata() throws E, IOException, InterruptedException;

    @NotNull FileSystemResource<E> getResource(@NotNull String id) throws E, IOException, InterruptedException;
    
    @NotNull ModuleDefinition getModule(@NotNull ModulePath id) throws E, IOException, InterruptedException;
    @NotNull ClassDefinition getClass(BigInteger id) throws E, IOException, InterruptedException;
    @NotNull TraitDefinition getTrait(BigInteger id) throws E, IOException, InterruptedException;
    @NotNull FunctionDefinition getFunction(BigInteger id) throws E, IOException, InterruptedException;
    @NotNull MethodDefinition getMethod(BigInteger id) throws E, IOException, InterruptedException;
    @NotNull ClassConstructorDefinition getClassConstructor(BigInteger id) throws E, IOException, InterruptedException;
}
